gpt4all-api/README.md

```diff
@@ -1,87 +1,158 @@
+
 # GPT4All REST API
+
 This directory contains the source code to run and build docker images that run a FastAPI app
 for serving inference from GPT4All models. The API matches the OpenAI API spec.
 
+The GPT4ALL REST API runs on port 4891 and supports CPU only. Edit the .env file to add the model, n_threads, embedding and USER_ID (instructions in /gpt4all/gpt4all-api/env). The location of the model is `gpt4all-api/gpt4all_api/models` directory.
+
 ## Tutorial
 
-The following tutorial assumes that you have checked out this repo and cd'd into it.
+This tutorial assumes that you have checked out this repo and cd'd into it. For any questions about docker, reference the docker documentation.
 
-### Starting the app
+### Building the backend API
 
-First change your working directory to `gpt4all/gpt4all-api`.
+Change your working directory to `gpt4all/gpt4all-api`.
+
+#### Option 1
+Build the FastAPI docker image. This is necessary on initial build or when adding new dependencies to the requirements.txt file. You need to specify USER_ID that exists in the host system:
 
-Now you can build the FastAPI docker image. You only have to do this on initial build or when you add new dependencies to the requirements.txt file:
 ```bash
-DOCKER_BUILDKIT=1 docker build -t gpt4all_api --progress plain -f gpt4all_api/Dockerfile.buildkit .
+DOCKER_BUILDKIT=1 docker build -t gpt4all_api --progress plain --build-arg USER_ID=1001 -f gpt4all_api/Dockerfile.buildkit .
 ```
 
-Then, start the backend with:
-
+You can then run docker compose to bring the container up, don't forget to edit the ENV file that contains varibles for the compose file
 ```bash
-docker compose up --build
+docker-compose up 
 ```
 
-This will run both the API and locally hosted GPU inference server. If you want to run the API without the GPU inference server, you can run:
+#### Option 2
+Alternatively, leverage the ENV file to set USER_ID and spin the container up, the "-d" will run it in the background
 
 ```bash
-docker compose up --build gpt4all_api
+docker compose up --build -d
```

Comment:
Running this command with the following `.env` file

```
MODEL_BIN="mistral-7b-openorca.Q4_0.gguf"
EMBEDDING="" # Optional, if you want to use a custom embedding, place it in the models folder and add the name here
N_THREADS=4 # 1 is default, 0 is all cores, this is an integer value not a string
USER_ID=1001 # 1001 is default, this is a string value
```

gives me the following error when POSTing chat completion requests:

> Process SpawnProcess-1:
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
>     self.run()
>   File "/usr/local/lib/python3.11/multiprocessing/process.py", line 108, in run
>     self._target(*self._args, **self._kwargs)
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/_subprocess.py", line 76, in subprocess_started
>     target(sockets=sockets)
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 59, in run
>     return asyncio.run(self.serve(sockets=sockets))
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/asyncio/runners.py", line 190, in run
>     return runner.run(main)
>            ^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/asyncio/runners.py", line 118, in run
>     return self._loop.run_until_complete(task)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 66, in serve
>     config.load()
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/config.py", line 471, in load
>     self.loaded_app = import_from_string(self.app)
>                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/importer.py", line 21, in import_from_string
>     module = importlib.import_module(module_str)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
>   File "/app/main.py", line 16, in <module>
>     from api_v1.api import router as v1_router
>   File "/app/api_v1/api.py", line 2, in <module>
>     from .routes import chat, completions, embeddings, engines, health
>   File "/app/api_v1/routes/embeddings.py", line 62, in <module>
>     embedder = Embed4All(model_path=settings.gpt4all_path)
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/appuser/.local/lib/python3.11/site-packages/gpt4all/gpt4all.py", line 44, in __init__
>     self.gpt4all = GPT4All(model_name or 'all-MiniLM-L6-v2-f16.gguf', n_threads=n_threads, **kwargs)
>                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/appuser/.local/lib/python3.11/site-packages/gpt4all/gpt4all.py", line 97, in __init__
>     self.config: ConfigType = self.retrieve_model(model_name, model_path=model_path, allow_download=allow_download, verbose=verbose)
>                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/appuser/.local/lib/python3.11/site-packages/gpt4all/gpt4all.py", line 187, in retrieve_model
>     config["path"] = GPT4All.download_model(model_filename, model_path, verbose=verbose, url=url)
>                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/appuser/.local/lib/python3.11/site-packages/gpt4all/gpt4all.py", line 239, in download_model
>     with open(download_path, "wb") as file, \
>          ^^^^^^^^^^^^^^^^^^^^^^^^^
> PermissionError: [Errno 13] Permission denied: '/models/all-MiniLM-L6-v2-f16.gguf'

Did I overlook some place where I should have changed "all-MiniLM-L6-v2-f16.gguf" to "mistral-7b-openorca.Q4_0.gguf"?

Reply:
```
-> % id
uid=1000(patrick) gid=1000(patrick) groups=1000(patrick),3(sys),90(network),98(power),108(vboxusers),967(docker),991(lp),998(wheel)

-> % ls -lad gpt4all_api/models 
drwxr-xr-x 2 patrick patrick 4096  5. Jan 09:19 gpt4all_api/models

-> % ls -lad gpt4all_api/models/mistral-7b-openorca.Q4_0.gguf 
-rw-r--r-- 1 patrick patrick 4108927744  5. Jan 09:19 gpt4all_api/models/mistral-7b-openorca.Q4_0.gguf
```

Reply:
Yes, I see the problem. In the instructions, I did document that you will to use the same posix ID as the host when building and running docker. The system doesn't care about the "appuser" but looks for the ID for read/write permissions on the folders which happen to be mounted from the host.

In short, use USER=1000

You could still use other users but then you need to adjust your project's posix read/write permissions on all folders to allow USER=1001 access.

Reply:
Thank you, this solves this issue!

Using the Python code in the README file does not give me a text back, though:
```
{
  "id": "6416e937-b289-4b4c-9fc2-e89ba12c1846",
  "object": "text_completion",
  "created": 1704462349,
  "model": "mistral-7b-openorca.Q4_0",
  "choices": [
    {
      "text": "\n",
      "index": 0,
      "logprobs": -1.0,
      "finish_reason": "length"
    }
  ],
  "usage": {
    "prompt_tokens": 22,
    "completion_tokens": 1,
    "total_tokens": 23
  }
}
```

The strange thing is that '"prompt_tokens": 22` seems to reflect each char in the input prompt "Who is Michael Jordan?". I'll see if I can figure it out.